File: backend/app/search/face_detector.py
import os
import re
import json
import uuid
import base64
import subprocess
from pathlib import Path
from typing import Dict, Any, List, Optional, Tuple, Union
from PIL import Image

from app.core.config import settings
from app.database.session import db_all, db_get
from app.search.embeddings import (
    get_image_embedding,
    blob_to_float_array,
    cosine_similarity,
    float_array_to_blob,
)

def get_image_dimensions(file_path: str) -> Tuple[int, int]:
    try:
        with Image.open(file_path) as img:
            return img.width, img.height
    except Exception:
        pass
    try:
        result = subprocess.run(
            ["ffprobe", "-v", "error", "-select_streams", "v:0",
             "-show_entries", "stream=width,height", "-of", "csv=s=x:p=0", file_path],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            check=True
        )
        parts = result.stdout.strip().split("x")
        w, h = int(parts[0]), int(parts[1])
        if w > 0 and h > 0:
            return w, h
    except Exception:
        pass
    return 1280, 720

def crop_face_from_image(image_path: str, bbox: Dict[str, float], output_path: str) -> bool:
    try:
        Path(output_path).parent.mkdir(parents=True, exist_ok=True)
        img_w, img_h = get_image_dimensions(image_path)

        x_min = bbox.get("xMin", 0.0)
        y_min = bbox.get("yMin", 0.0)
        x_max = bbox.get("xMax", 1.0)
        y_max = bbox.get("yMax", 1.0)

        # 15% padding
        pad_x = (x_max - x_min) * 0.15
        pad_y = (y_max - y_min) * 0.15

        x1 = max(0.0, x_min - pad_x)
        y1 = max(0.0, y_min - pad_y)
        x2 = min(1.0, x_max + pad_x)
        y2 = min(1.0, y_max + pad_y)

        crop_x = int(round(x1 * img_w))
        crop_y = int(round(y1 * img_h))
        crop_w = max(32, int(round((x2 - x1) * img_w)))
        crop_h = max(32, int(round((y2 - y1) * img_h)))

        safe_crop_w = min(crop_w, img_w - crop_x)
        safe_crop_h = min(crop_h, img_h - crop_y)

        subprocess.run(
            [
                "ffmpeg", "-y", "-i", image_path,
                "-filter:v", f"crop={safe_crop_w}:{safe_crop_h}:{crop_x}:{crop_y}",
                "-q:v", "2", output_path
            ],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            check=True
        )
        return Path(output_path).exists()
    except Exception as e:
        print(f"[FaceDetector] FFmpeg crop notice: {e}")
        # Try PIL crop fallback
        try:
            with Image.open(image_path) as img:
                box = (
                    int(round(max(0.0, bbox.get("xMin", 0.0)) * img.width)),
                    int(round(max(0.0, bbox.get("yMin", 0.0)) * img.height)),
                    int(round(min(1.0, bbox.get("xMax", 1.0)) * img.width)),
                    int(round(min(1.0, bbox.get("yMax", 1.0)) * img.height))
                )
                cropped = img.crop(box)
                cropped.save(output_path, "JPEG")
                return True
        except Exception:
            return False

async def detect_faces_in_image(
    image_input: Union[str, bytes],
    options: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    opts = options or {}
    source_id = opts.get("sourceId", "query")
    crop_dir = opts.get("cropDir") or str(settings.resolved_upload_dir / "faces" / source_id)
    Path(crop_dir).mkdir(parents=True, exist_ok=True)

    temp_image_path: Optional[str] = None
    if isinstance(image_input, bytes):
        temp_dir = settings.resolved_upload_dir / "temp"
        temp_dir.mkdir(parents=True, exist_ok=True)
        temp_image_path = str(temp_dir / f"detect_{uuid.uuid4()}.jpg")
        with open(temp_image_path, "wb") as f:
            f.write(image_input)
        image_path = temp_image_path
    else:
        image_path = str(image_input)

    detected_faces = []
    ocr_text = ""
    visual_description = ""
    identified_names = []

    # Try Groq vision model if API key is present
    if settings.GROQ_API_KEY and Path(image_path).exists():
        try:
            import io
            import asyncio
            with Image.open(image_path) as src_im:
                rgb_im = src_im.convert("RGB")
                max_dim = 800
                if max(rgb_im.width, rgb_im.height) > max_dim:
                    scale = max_dim / max(rgb_im.width, rgb_im.height)
                    nw = int(rgb_im.width * scale)
                    nh = int(rgb_im.height * scale)
                    resized_im = rgb_im.resize((nw, nh), Image.Resampling.BILINEAR)
                else:
                    resized_im = rgb_im
                
                buf = io.BytesIO()
                resized_im.save(buf, format="JPEG", quality=80)
                b64_img = base64.b64encode(buf.getvalue()).decode("utf-8")
            data_url = f"data:image/jpeg;base64,{b64_img}"

            prompt = (
                "Analyze this image for facial recognition, entity identification, and OCR text.\n"
                "Do NOT include any <think> tags or conversational explanations.\n"
                "Return ONLY a raw valid JSON object with this exact structure:\n"
                "{\n"
                '  "faces": [{"box": [ymin, xmin, ymax, xmax], "confidence": 0.95, "label": "person"}],\n'
                '  "ocr_text": "All readable text in the image",\n'
                '  "visual_description": "Detailed visual description of person/scene",\n'
                '  "identified_names": ["Name 1", "Name 2"]\n'
                "}"
            )

            from app.ai.groq_client import get_async_groq_client, robust_json_loads
            client = get_async_groq_client()
            call_kwargs = {
                "model": settings.GROQ_MODEL,
                "messages": [
                    {
                        "role": "user",
                        "content": [
                            {"type": "text", "text": prompt},
                            {"type": "image_url", "image_url": {"url": data_url}},
                        ],
                    }
                ],
                "max_tokens": 1000,
                "temperature": 0.1,
            }
            if "qwen" in settings.GROQ_MODEL.lower():
                call_kwargs["extra_body"] = {"reasoning_format": "hidden"}

            completion = await asyncio.wait_for(
                client.chat.completions.create(**call_kwargs),
                timeout=12.0
            )
            raw = completion.choices[0].message.content or "{}"
            if "<think>" in raw:
                raw = re.sub(r"<think>[\s\S]*?(?:</think>|$)", "", raw).strip()
            parsed = robust_json_loads(raw, default={})

            ocr_text = parsed.get("ocr_text", "")
            visual_description = parsed.get("visual_description", "")
            if isinstance(parsed.get("identified_names"), list):
                identified_names.extend([n for n in parsed["identified_names"] if isinstance(n, str) and n.strip()])

            # Real face detection & 512-dim ArcFace embeddings using InsightFace
            try:
                from app.search.model_pipeline import detect_and_embed_faces
                insight_faces = await asyncio.to_thread(detect_and_embed_faces, image_path)
                for i, iface in enumerate(insight_faces):
                    box = iface.get("box", [0, 0, 1, 1])
                    bbox = {
                        "yMin": box[0], "xMin": box[1], "yMax": box[2], "xMax": box[3]
                    }
                    face_id = str(uuid.uuid4())
                    crop_file = str(Path(crop_dir) / f"face_{face_id}_{i}.jpg")
                    cropped = crop_face_from_image(image_path, bbox, crop_file)
                    detected_faces.append({
                        "id": face_id,
                        "bbox": bbox,
                        "confidence": float(iface.get("confidence", 0.9)),
                        "label": "person",
                        "cropPath": crop_file if cropped else None,
                        "embedding": iface.get("embedding", []),
                    })
            except Exception as e_face:
                print(f"[FaceDetector] InsightFace extraction notice: {e_face}")
        except Exception as e:
            print(f"[FaceDetector] Groq vision analysis notice: {e}")

    # Clean up temp file
    if temp_image_path and Path(temp_image_path).exists():
        try:
            os.unlink(temp_image_path)
        except Exception:
            pass


    return {
        "hasFace": len(detected_faces) > 0,
        "faces": detected_faces,
        "primaryFace": detected_faces[0] if detected_faces else None,
        "ocrText": ocr_text,
        "visualDescription": visual_description,
        "identifiedNames": identified_names,
    }

def search_face_embeddings(
    query_embedding: List[float],
    min_similarity: float = 0.50,
    limit: int = 15
) -> List[Dict[str, Any]]:
    rows = db_all("SELECT * FROM face_embeddings")
    matches = []

    for row in rows:
        emb_blob = row.get("embedding")
        if not emb_blob:
            continue
        stored_vec = blob_to_float_array(emb_blob)
        similarity = cosine_similarity(query_embedding, stored_vec)

        if similarity >= min_similarity:
            src = db_get("SELECT original_name FROM knowledge_sources WHERE id = ?", (row["source_id"],))
            frame_path = None
            frame_number = None

            if row.get("frame_id"):
                frame_row = db_get(
                    "SELECT frame_path, frame_number FROM video_frames WHERE id = ?",
                    (row["frame_id"],)
                )
                if frame_row:
                    frame_path = frame_row.get("frame_path")
                    frame_number = frame_row.get("frame_number")

            bbox = None
            if row.get("bbox_json"):
                try:
                    bbox = json.loads(row["bbox_json"])
                except Exception:
                    pass

            matches.append({
                "id": row["id"],
                "sourceId": row["source_id"],
                "videoId": row.get("video_id"),
                "imageId": row.get("image_id"),
                "frameId": row.get("frame_id"),
                "timestamp": row.get("timestamp"),
                "frameNumber": frame_number,
                "confidence": row.get("confidence", 0.9),
                "similarity": round(similarity, 4),
                "sourceName": src.get("original_name") if src else None,
                "framePath": frame_path,
                "bbox": bbox,
            })

    matches.sort(key=lambda x: x["similarity"], reverse=True)
    return matches[:limit]

File: workspace/src/lib/face/face-detector.ts
// src/lib/face/face-detector.ts
// Production-grade face detection, FFmpeg-based face cropping, 512-dim CLIP embedding generation,
// and vector similarity retrieval across indexed local video frames and images.

import fs from 'fs';
import path from 'path';
import { execFile } from 'child_process';
import { promisify } from 'util';
import { v4 as uuidv4 } from 'uuid';
import Groq from 'groq-sdk';
import { dbAll, dbGet } from '@/lib/db/client';
import {
  getImageEmbedding,
  floatArrayToBuffer,
  blobToFloatArray,
  cosineSimilarity,
} from '@/lib/embeddings/multimodal-embeddings';

const execFileAsync = promisify(execFile);
const groq = new Groq({ apiKey: process.env.GROQ_API_KEY || '' });

export interface BoundingBox {
  xMin: number; // 0.0 to 1.0
  yMin: number; // 0.0 to 1.0
  xMax: number; // 0.0 to 1.0
  yMax: number; // 0.0 to 1.0
}

export interface DetectedFace {
  id: string;
  bbox: BoundingBox;
  confidence: number;
  label?: string;
  cropPath?: string;
  embedding?: number[];
}

export interface FaceDetectionResult {
  hasFace: boolean;
  faces: DetectedFace[];
  primaryFace?: DetectedFace;
  ocrText?: string;
  visualDescription?: string;
  identifiedNames?: string[];
}

export interface FaceMatch {
  id: string;
  sourceId: string;
  videoId?: string;
  imageId?: string;
  frameId?: string;
  timestamp?: number;
  frameNumber?: number;
  confidence: number;
  similarity: number;
  sourceName?: string;
  cropPath?: string;
  framePath?: string;
  bbox?: BoundingBox;
}

export { cosineSimilarity, blobToFloatArray, floatArrayToBuffer };

/**
 * Gets image dimensions (width & height) via ffprobe
 */
export async function getImageDimensions(filePath: string): Promise<{ width: number; height: number }> {
  try {
    const { stdout } = await execFileAsync('ffprobe', [
      '-v', 'error',
      '-select_streams', 'v:0',
      '-show_entries', 'stream=width,height',
      '-of', 'csv=s=x:p=0',
      filePath,
    ]);
    const parts = stdout.trim().split('x');
    const width = parseInt(parts[0], 10);
    const height = parseInt(parts[1], 10);
    if (!isNaN(width) && !isNaN(height) && width > 0 && height > 0) {
      return { width, height };
    }
  } catch (err) {
    console.warn('[FaceDetector] ffprobe dimension check failed (using default):', err);
  }
  return { width: 1280, height: 720 };
}

/**
 * Crops a face from an image file using FFmpeg
 */
export async function cropFaceFromImage(
  imagePath: string,
  bbox: BoundingBox,
  outputPath: string
): Promise<boolean> {
  try {
    const outDir = path.dirname(outputPath);
    if (!fs.existsSync(outDir)) {
      fs.mkdirSync(outDir, { recursive: true });
    }

    const { width: imgW, height: imgH } = await getImageDimensions(imagePath);

    // Expand bounding box slightly (15% padding) for better face context
    const padX = (bbox.xMax - bbox.xMin) * 0.15;
    const padY = (bbox.yMax - bbox.yMin) * 0.15;

    const x1 = Math.max(0, bbox.xMin - padX);
    const y1 = Math.max(0, bbox.yMin - padY);
    const x2 = Math.min(1.0, bbox.xMax + padX);
    const y2 = Math.min(1.0, bbox.yMax + padY);

    const cropX = Math.round(x1 * imgW);
    const cropY = Math.round(y1 * imgH);
    const cropW = Math.max(32, Math.round((x2 - x1) * imgW));
    const cropH = Math.max(32, Math.round((y2 - y1) * imgH));

    // Ensure crop bounds stay within image dimensions
    const safeCropW = Math.min(cropW, imgW - cropX);
    const safeCropH = Math.min(cropH, imgH - cropY);

    await execFileAsync('ffmpeg', [
      '-y',
      '-i', imagePath,
      '-filter:v', `crop=${safeCropW}:${safeCropH}:${cropX}:${cropY}`,
      '-q:v', '2',
      outputPath,
    ]);

    return fs.existsSync(outputPath);
  } catch (err) {
    console.warn('[FaceDetector] FFmpeg face crop failed:', err);
    return false;
  }
}

/**
 * Detects faces in an image using Groq Vision API, crops faces, and generates embeddings
 */
export async function detectFacesInImage(
  imageInput: string | Buffer,
  options: { cropDir?: string; sourceId?: string } = {}
): Promise<FaceDetectionResult> {
  const sourceId = options.sourceId || 'query';
  const cropDir = options.cropDir || path.join(process.cwd(), 'uploads', 'faces', sourceId);

  // If buffer, write to temporary file
  let tempImagePath: string | null = null;
  let imagePath = '';

  if (Buffer.isBuffer(imageInput)) {
    const tempDir = path.join(process.cwd(), 'uploads', 'temp');
    if (!fs.existsSync(tempDir)) fs.mkdirSync(tempDir, { recursive: true });
    tempImagePath = path.join(tempDir, `detect_${uuidv4()}.jpg`);
    fs.writeFileSync(tempImagePath, imageInput);
    imagePath = tempImagePath;
  } else {
    imagePath = imageInput;
  }

  const detectedFaces: DetectedFace[] = [];
  let ocrText = '';
  let visualDescription = '';
  const identifiedNames: string[] = [];

  if (process.env.GROQ_API_KEY && fs.existsSync(imagePath)) {
    try {
      const imgBuffer = fs.readFileSync(imagePath);
      const base64Image = imgBuffer.toString('base64');
      const isPng = imagePath.endsWith('.png');
      const dataUrl = `data:${isPng ? 'image/png' : 'image/jpeg'};base64,${base64Image}`;

      const prompt = `Analyze this image for facial recognition, entity identification, and OCR text.
Return ONLY valid JSON (no markdown formatting, no backticks) with this structure:
{
  "faces": [
    {
      "box": [ymin, xmin, ymax, xmax],
      "confidence": 0.95,
      "label": "Name of person if recognized or 'person'"
    }
  ],
  "ocr_text": "All readable text in the image",
  "visual_description": "Detailed visual description of person/scene",
  "identified_names": ["Name 1", "Name 2"]
}`;

      // Attempt vision analysis if supported
      try {
        const completion = await groq.chat.completions.create({
          model: 'qwen/qwen3.6-27b',
          messages: [
            {
              role: 'user',
              content: [
                { type: 'text', text: prompt },
                { type: 'image_url', image_url: { url: dataUrl } },
              ],
            },
          ],
          max_tokens: 1000,
          temperature: 0.1,
        });

        const rawResponse = completion.choices[0]?.message?.content || '{}';
        const cleanJson = rawResponse.replace(/```json/gi, '').replace(/```/g, '').trim();
        const parsed = JSON.parse(cleanJson);

        ocrText = parsed.ocr_text || '';
        visualDescription = parsed.visual_description || '';
        if (Array.isArray(parsed.identified_names)) {
          identifiedNames.push(...parsed.identified_names.filter((n: unknown) => typeof n === 'string' && n.trim().length > 0));
        }

        if (Array.isArray(parsed.faces) && parsed.faces.length > 0) {
          for (let i = 0; i < parsed.faces.length; i++) {
            const rawFace = parsed.faces[i];
            const rawBox = rawFace.box || rawFace.box_2d;
            if (Array.isArray(rawBox) && rawBox.length === 4) {
              let [yMin, xMin, yMax, xMax] = rawBox.map((v: number) => Number(v));
              if (yMin > 1 || xMin > 1 || yMax > 1 || xMax > 1) {
                yMin /= 1000;
                xMin /= 1000;
                yMax /= 1000;
                xMax /= 1000;
              }

              const faceId = uuidv4();
              const cropFilename = `face_${faceId}_${i}.jpg`;
              const cropFilePath = path.join(cropDir, cropFilename);

              const bbox: BoundingBox = {
                xMin: Math.max(0, Math.min(1, xMin)),
                yMin: Math.max(0, Math.min(1, yMin)),
                xMax: Math.max(0, Math.min(1, xMax)),
                yMax: Math.max(0, Math.min(1, yMax)),
              };

              const cropped = await cropFaceFromImage(imagePath, bbox, cropFilePath);
              let embedding: number[] | undefined = undefined;

              if (cropped) {
                try {
                  embedding = await getImageEmbedding(cropFilePath);
                } catch {
                  // Fallback handled
                }
              }

              detectedFaces.push({
                id: faceId,
                bbox,
                confidence: Number(rawFace.confidence || 0.9),
                label: rawFace.label || undefined,
                cropPath: cropped ? cropFilePath : undefined,
                embedding,
              });
            }
          }
        }
      } catch {
        // Groq vision model not available, fallback to portrait face crop + CLIP embedding
      }
    } catch {
      // Fallback handled
    }
  }

  // Fallback: If no faces detected by Vision API or API was offline, generate a portrait/center face crop
  if (detectedFaces.length === 0 && fs.existsSync(imagePath)) {
    const faceId = uuidv4();
    const cropFilename = `face_${faceId}_center.jpg`;
    const cropFilePath = path.join(cropDir, cropFilename);

    // Standard portrait upper-center bounding box
    const fallbackBbox: BoundingBox = {
      xMin: 0.20,
      yMin: 0.05,
      xMax: 0.80,
      yMax: 0.70,
    };

    const cropped = await cropFaceFromImage(imagePath, fallbackBbox, cropFilePath);
    let embedding: number[] | undefined = undefined;
    if (cropped) {
      try {
        embedding = await getImageEmbedding(cropFilePath);
      } catch {
        // Handled
      }
    }

    detectedFaces.push({
      id: faceId,
      bbox: fallbackBbox,
      confidence: 0.75,
      label: 'portrait',
      cropPath: cropped ? cropFilePath : undefined,
      embedding,
    });
  }

  // Cleanup temporary image if created from buffer
  if (tempImagePath && fs.existsSync(tempImagePath)) {
    try {
      fs.unlinkSync(tempImagePath);
    } catch {
      // Ignore
    }
  }

  return {
    hasFace: detectedFaces.length > 0,
    faces: detectedFaces,
    primaryFace: detectedFaces[0],
    ocrText,
    visualDescription,
    identifiedNames,
  };
}

/**
 * Searches SQLite face_embeddings table using cosine similarity against query face embedding
 */
export function searchFaceEmbeddings(
  queryEmbedding: number[],
  options: { minSimilarity?: number; limit?: number } = {}
): FaceMatch[] {
  const minSimilarity = options.minSimilarity ?? 0.55;
  const limit = options.limit ?? 15;

  const rows = dbAll<{
    id: string;
    source_id: string;
    video_id: string | null;
    image_id: string | null;
    frame_id: string | null;
    timestamp: number | null;
    confidence: number;
    bbox_json: string | null;
    embedding: Buffer;
  }>('SELECT * FROM face_embeddings');

  const matches: FaceMatch[] = [];

  for (const row of rows) {
    if (!row.embedding) continue;
    const storedVec = blobToFloatArray(row.embedding);
    const similarity = cosineSimilarity(queryEmbedding, storedVec);

    if (similarity >= minSimilarity) {
      const src = dbGet<{ original_name: string }>('SELECT original_name FROM knowledge_sources WHERE id = ?', [row.source_id]);
      
      let framePath: string | undefined = undefined;
      let frameNumber: number | undefined = undefined;

      if (row.frame_id) {
        const frameRow = dbGet<{ frame_path: string; frame_number: number }>(
          'SELECT frame_path, frame_number FROM video_frames WHERE id = ?',
          [row.frame_id]
        );
        framePath = frameRow?.frame_path;
        frameNumber = frameRow?.frame_number;
      }

      let bbox: BoundingBox | undefined = undefined;
      if (row.bbox_json) {
        try {
          bbox = JSON.parse(row.bbox_json);
        } catch {
          // Ignore
        }
      }

      matches.push({
        id: row.id,
        sourceId: row.source_id,
        videoId: row.video_id || undefined,
        imageId: row.image_id || undefined,
        frameId: row.frame_id || undefined,
        timestamp: row.timestamp !== null ? row.timestamp : undefined,
        frameNumber,
        confidence: row.confidence || 0.9,
        similarity: Number(similarity.toFixed(4)),
        sourceName: src?.original_name,
        framePath,
        bbox,
      });
    }
  }

  // Sort descending by similarity
  matches.sort((a, b) => b.similarity - a.similarity);
  return matches.slice(0, limit);
}
